### PR TITLE
Add warning about locally closed resources being undefined behavior

### LINKED ISSUE #106: Streams closed before watcher is enabled
Since read / write watchers are enabled after the current tick of the event loop there is a possibility that a stream resource is closed by another watcher within the same tick. This causes trouble with many loop backends (ev, event, uv) because they fail at creating new event watchers when the stream resource has been closed.

This is an edge case but i think that the specification should be updated to define expected behavior in this case. It would be sufficient to have loop implementations check for valid resources as new stream watchers are enabled for the first time and call the watcher callback once if a closed / invalid resource is detected.

### PR DESCRIPTION
**Pull request by kelunik**
Also makes Loop docs consistent with Driver. Resolves #106.

### CODE REVIEW

File: src/Loop.php [hunk 1]
```diff
@@ -178,11 +178,18 @@ public static function repeat($interval, callable $callback, $data = null)
     /**
      * Execute a callback when a stream resource becomes readable or is closed for reading.
      *
+     * Warning: Closing resources locally, e.g. with `fclose`, might not invoke the callback. Be sure to `cancel` the
+     * watcher when closing the resource locally. Drivers might choose to notify the user in a debug mode if there are
```
Comment (bwoebi) [inline src/Loop.php]: s/might choose to notify the user in a debug mode/may choose to notify the user

There's no need for talking about a debug mode - that's an implementation detail unrelated to the specification here.

**Review by bwoebi — CHANGES_REQUESTED** [review]

File: src/Loop.php (hunk 1 above)
Comment (kelunik) [reply src/Loop.php]: There's no _need_, but just a hint.

Comment (bwoebi) [reply src/Loop.php]: The emphasis is on removing "in a debug mode". Just keep it out, it has no business here.
Maybe put it in a meta document outlining how you intend it to be used, but it's not appropriate here.

**Review by bwoebi — APPROVED** [review]

**Review by WyriHaximus — APPROVED** [review]
:+1:

File: src/Loop/Driver.php [hunk 2]
```diff
@@ -81,9 +81,9 @@
      * Execute a callback when a stream resource becomes readable or is closed for reading.
      *
      * Warning: Closing resources locally, e.g. with `fclose`, might not invoke the callback. Be sure to `cancel` the
-     * watcher when closing the resource locally. Drivers might choose to notify the user in a debug mode if there are
-     * watchers on invalid resources, but are not required to, due to the high performance impact. Watchers on closed
-     * resources are therefore undefined behavior.
+     * watcher when closing the resource locally. Drivers may choose to notify the user if there are watchers on invalid
```
Comment (kelunik) [inline src/Loop/Driver.php]: Maybe we should write `MAY`

**Comment by kelunik** [comment]
Forgot to copy `defer`, `delay`, `repeat` from `Driver` to `Loop`. Also replaced `may` with `MAY` and rebased onto `master`.

**Comment by bwoebi** [comment]
Feel free - would be consistent with the other text.

Bob Weinand (iPhone)

**Comment by kelunik** [comment]
That's a bit late.

**Pull request merged**